### PR TITLE
Handle 0 rows

### PR DESCRIPTION
**Pull request by OliverJAsh**
# Problem

See the failing test. Previously this would throw an exception:

```
  ● BuildComparator › toMarkdown › does not error when there are 0 rows

    TypeError: Cannot read property 'map' of undefined

      412 |     const rows = this.getStringFormattedRows(formatTotal, formatDelta, sizeKey, artifactFilter);
      413 |
    > 414 |     return markdownTable([header, ...groups, ...rows], { align: rows[0].map((_, i) => (i === 0 ? 'l' : 'r')) });
          |                                                                         ^
      415 |   }
      416 |
      417 |   public toCsv({

      at BuildComparator.Object.<anonymous>.BuildComparator.toMarkdown (src/index.ts:414:73)
      at Object.<anonymous> (src/__tests__/index.test.ts:351:25)
```

# Solution

Explain your approach. Sometimes it helps to justify your approach against some others that you didn't choose to explain why yours is better.

# TODO

- [x] 🤓 Add & update tests (always try to _increase_ test coverage)
- [x] 🔬 Ensure CI is passing (`yarn lint:ci`, `yarn test`, `yarn tsc:ci`)
- [x] 📖 Update relevant documentation

### CODE REVIEW

File: src/comparator/src/index.ts [hunk 1]
```diff
@@ -411,7 +411,9 @@ export default class BuildComparator {
     const groups = this.getStringFormattedGroups(formatTotal, formatDelta, sizeKey);
     const rows = this.getStringFormattedRows(formatTotal, formatDelta, sizeKey, artifactFilter);
 
-    return markdownTable([header, ...groups, ...rows], { align: rows[0].map((_, i) => (i === 0 ? 'l' : 'r')) });
```
Comment (OliverJAsh) [inline src/comparator/src/index.ts]: There are a number of other places where we're assuming the array will always be non-empty (i.e. `xs[0].map(…)`), which might suffer from similar bugs if it turns out the array might in fact be empty.

Comment (paularmstrong) [reply src/comparator/src/index.ts]: A build shouldn't be allowed in by the API that has less than one artifact. Is there another case where this is possible?

Comment (montogeek) [reply src/comparator/src/index.ts]: Not possible to use optional chaining here?

Comment (OliverJAsh) [reply src/comparator/src/index.ts]: > Is there another case where this is possible?

I believe so—see the test I added. `artifactFilter` may return false for all items. This is actually what happened to me when testing the example used here: https://buildtracker.dev/docs/guides/ci

> Not possible to use optional chaining here?

Sure, I'll give it a shot!

Comment (paularmstrong) [reply src/comparator/src/index.ts]: Ah filters, of course!

I didn't do anything to enable optional chaining—not sure if it's enabled by default. Happy to start including it, though!

Comment (OliverJAsh) [reply src/comparator/src/index.ts]: I'm not sure the optional chaining operator can help here, but I tried to use nullish coalescing. Unfortunately it seems the version of ESLint that is currently used does not have support:

image

```sh-session
❯ npm run lint

> build-tracker@1.0.0 lint /Users/oliverash/Development/build-tracker
> yarn lint:cmd --fix

yarn run v1.19.1
$ eslint . --ext '.ts,.tsx,.js,.json,.md,.snap' --cache --fix
{ parser: "babylon" } is deprecated; we now treat it as { parser: "babel" }.

/Users/oliverash/Development/build-tracker/src/comparator/src/index.ts
  414:30  error  Parsing error: Expression expected
```

Comment (paularmstrong) [reply src/comparator/src/index.ts]: I plan on upgrading eslint, typescript, and a few other things today

File: src/comparator/src/index.ts [hunk 2]
```diff
@@ -411,7 +411,9 @@ export default class BuildComparator {
...
+    const firstRow = rows[0] !== undefined ? rows[0] : [];
+
+    return markdownTable([header, ...groups, ...rows], { align: firstRow.map((_, i) => (i === 0 ? 'l' : 'r')) });
```
Comment (paularmstrong) [inline src/comparator/src/index.ts]: Looking at this all again, it looks like we lose the alignment in the columns. I think we could avoid the check for row length via:

```ts
const allRows = [header, ...groups, ...rows];
return markdownTable(allRows, { align: allRows[0].map((_, i) => (i === 0 ? 'l' : 'r')) });
```

Comment (OliverJAsh) [reply src/comparator/src/index.ts]: Good shout!

**Pull request merged**